File: src/main/java/jk/kamoru/ground/flay/service/ScoreCalculator.java
package jk.kamoru.ground.flay.service;

import java.util.Collection;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import jk.kamoru.ground.GroundProperties;
import jk.kamoru.ground.flay.domain.Flay;
import jk.kamoru.ground.info.domain.Actress;
import jk.kamoru.ground.info.domain.Tag;
import jk.kamoru.ground.info.source.InfoSource;

@Component
public class ScoreCalculator {

  @Autowired
  GroundProperties properties;

  @Autowired
  InfoSource<Actress, String> actressInfoSource;

  Comparator<Flay> likeComparator = Comparator.comparing(Flay::getLikeCount);
  Comparator<Flay> scoreComparator = Comparator.comparing(Flay::getScore);
  Comparator<Flay> studioPointComparator = Comparator.comparing(Flay::getStudioPoint);
  Comparator<Flay> actressPointComparator = Comparator.comparing(Flay::getActressPoint);
  Comparator<Flay> modifiedComparator = Comparator.comparing(Flay::getLastModified);
  Comparator<Flay> releaseComparator = Comparator.comparing(Flay::getRelease);

  Map<String, AtomicInteger> studioCountMap = new HashMap<>();

  public Collection<Flay> listOrderByScoreDesc(Collection<Flay> flayList) {
    // rank 0 제외, 비디오가 없는 것 제외
    List<Flay> filteredList = flayList.stream().filter(flay -> flay.getVideo().getRank() > 0 && flay.getFiles().get(Flay.MOVIE).size() > 0).toList();

    filteredList.forEach(f -> {
      if (studioCountMap.containsKey(f.getStudio())) {
        studioCountMap.get(f.getStudio()).incrementAndGet();
      } else {
        studioCountMap.put(f.getStudio(), new AtomicInteger(1));
      }
    });

    filteredList.forEach(f -> {
      calcScore(f);
      setActressPoint(f, filteredList);
      f.setStudioPoint(studioCountMap.get(f.getStudio()).intValue());
    });

    return filteredList.stream()
        .sorted(
          likeComparator.reversed().thenComparing(
            scoreComparator.reversed().thenComparing(
              studioPointComparator.reversed().thenComparing(
                actressPointComparator.reversed().thenComparing(
                  modifiedComparator.reversed().thenComparing(
                    releaseComparator.reversed())))))).toList();
  }

  public void calcScore(Flay flay) {
    flay.setScore(
        resolveLikes(flay) * properties.getScore().getLikePoint() +
        resolveRank(flay) * properties.getScore().getRankPoint() +
        // + flay.getVideo().getPlay() * properties.getScore().getPlayPoint()
        existingCountOfSubtitle(flay) * properties.getScore().getSubtitlesPoint() +
        countFavoriteActress(flay) * properties.getScore().getFavoritePoint());
  }

  private int resolveLikes(Flay flay) {
    return flay.getVideo().getLikes() == null ? 0 : flay.getVideo().getLikes().size();
  }

  /**
   * flay의 rank. rank == 0 이면, tag에서 찾는다
   *
   * @param flay
   * @return
   */
  private int resolveRank(Flay flay) {
    if (flay.getVideo().getRank() != 0) {
      return flay.getVideo().getRank();
    } else {
      // {1: 50, 2: 63, 3: 64, 4: 65, 5: 66}
      List<Integer> tags = flay.getVideo().getTags().stream().map(Tag::getId).toList();
      return tags.contains(66) ? 5 : tags.contains(65) ? 4 : tags.contains(64) ? 3 : tags.contains(63) ? 2 : tags.contains(50) ? 1 : 0;
    }
  }

  /**
   * 자막이 있으면 1, 없으면 0
   *
   * @param flay
   * @return
   */
  private int existingCountOfSubtitle(Flay flay) {
    return flay.getFiles().get(Flay.SUBTI).size() > 0 ? 1 : 0;
  }

  /**
   * flay의 배우들의 favorite 갯수
   *
   * @param flay
   * @return
   */
  private int countFavoriteActress(Flay flay) {
    return flay.getActressList().stream().mapToInt(a -> StringUtils.isNotBlank(a) && actressInfoSource.get(a).isFavorite() ? 1 : 0).sum();
  }

  // 출연배우의 작품 함계
  private void setActressPoint(Flay flay, Collection<Flay> flayList) {
    int count = 0;
    for (String name : flay.getActressList()) {
      for (Flay f : flayList) {
        if (f.getActressList().contains(name)) {
          count++;
        }
      }
    }
    flay.setActressPoint(count);
  }

}

File: web-backend/src/services/score-calculator.ts
import { config } from '../config';
import { Flay } from '../domain/flay';
import { actressInfoSource } from '../sources/info-sources';

/**
 * Flay 점수를 계산한다.
 * Java ScoreCalculator.calcScore() 대응
 *
 * score = likes * likePoint + rank * rankPoint + subtitle * subtitlesPoint + favoriteActress * favoritePoint
 */
export function calcScore(flay: Flay): void {
  const { rankPoint, likePoint, subtitlesPoint, favoritePoint } = config.flay.score;

  flay.score = resolveLikes(flay) * likePoint + resolveRank(flay) * rankPoint + existingCountOfSubtitle(flay) * subtitlesPoint + countFavoriteActress(flay) * favoritePoint;
}

/** likes 개수 */
function resolveLikes(flay: Flay): number {
  return flay.video.likes ? flay.video.likes.length : 0;
}

/**
 * rank 값을 구한다. 0이면 태그에서 추출한다.
 * {1: 50, 2: 63, 3: 64, 4: 65, 5: 66}
 */
function resolveRank(flay: Flay): number {
  if (flay.video.rank !== 0) return flay.video.rank;

  const tagIds = flay.video.tags.map((t) => t.id);
  if (tagIds.includes(66)) return 5;
  if (tagIds.includes(65)) return 4;
  if (tagIds.includes(64)) return 3;
  if (tagIds.includes(63)) return 2;
  if (tagIds.includes(50)) return 1;
  return 0;
}

/** 자막 존재 여부 (있으면 1, 없으면 0) */
function existingCountOfSubtitle(flay: Flay): number {
  return flay.files.subtitles.length > 0 ? 1 : 0;
}

/** 즐겨찾기 배우 수 */
function countFavoriteActress(flay: Flay): number {
  return flay.actressList.reduce((sum, name) => {
    if (!name) return sum;
    try {
      const actress = actressInfoSource.get(name);
      return sum + (actress.favorite ? 1 : 0);
    } catch {
      return sum;
    }
  }, 0);
}

/**
 * likes, score, release, modified 순으로 정렬된 리스트를 반환한다.
 * rank > 0 이고 영상 파일이 있는 것만 포함한다.
 * Java ScoreCalculator.listOrderByScoreDesc() 대응
 */
export function listOrderByScoreDesc(flayList: Flay[]): Flay[] {
  const filtered = flayList.filter((flay) => flay.video.rank > 0 && flay.files.movie.length > 0);

  filtered.forEach((f) => calcScore(f));

  return filtered.sort((a, b) => {
    // likes 내림차순
    const likeDiff = getLikeCount(b) - getLikeCount(a);
    if (likeDiff !== 0) return likeDiff;

    // score 내림차순
    const scoreDiff = b.score - a.score;
    if (scoreDiff !== 0) return scoreDiff;

    // release 내림차순
    const releaseDiff = b.release.localeCompare(a.release);
    if (releaseDiff !== 0) return releaseDiff;

    // lastModified 내림차순
    return b.lastModified - a.lastModified;
  });
}

/** likes 개수 */
function getLikeCount(flay: Flay): number {
  return flay.video.likes ? flay.video.likes.length : 0;
}
